python/paddle/fluid/contrib/int8_inference/utility.py

```diff
@@ -0,0 +1,731 @@
+                max_range = Calibrator.s8_max
+            else:
+                if i.name in self._conv_output_var_name:
+                    cur_op = self.program.current_block().ops[
+                        self._conv_output_var_name.index(i.name) + 2]
+                    if cur_op.has_attr('fuse_relu') and cur_op.attr(
+                            'fuse_relu'):
+                        max_range = Calibrator.u8_max
+                        self._u8_output_var.append(i.name)
+                    else:
+                        max_range = Calibrator.s8_max
+                        self._s8_output_var.append(i.name)
+                else:
+                    max_range = self.__get_max_range_by_var_name(self.program,
+                                                                 i.name)
+
+                max_value = [[np.abs(np_data)]
+                             for np_data in sampling_data[i.name]]
+
+            self._var_max_range[i.name] = max_range
+
+            self._var_max_value_map[i.name] = max_value
+
+    def __check_force_fp32_attr_by_output_var(self, program, var_name):
+        for op in program.current_block().ops:
+            if op.type == "conv2d" and var_name in op.output_arg_names:
+                return op.attr("force_fp32_output")
+        return False
+
+    def __get_op_index_by_output_var(self, program, var_name, start_index=0):
+        '''
+        Check the specified input variable is the output of the conv/pool2d
+        op's output or not.
+
+        Returns:
+            The index if the variable is the output of any conv/pool2d op's
+            output.
+            -1 when the variable is not the output of any conv/pool2d op's 
+            output.
+        '''
+        for index, op in enumerate(program.current_block().ops[start_index:]):
+            if var_name in op.output_arg_names and op.type in Calibrator.supported_int8_op_type:
+                return index
+        return -1
+
+    def __get_op_index_by_input_var(self, program, var_name, start_index=0):
+        '''
+        Get the op index by specified input variable.
+        Returns:
+            The op index if the variable is the input of this op or -1 if the 
+            variable is not the input of any op. 
+        '''
+        for index, op in enumerate(program.current_block().ops[start_index:]):
+            if var_name in op.input_arg_names:
+                return index
+
+        return -1
+
+    def __get_quantize_dequantize_combination(self, program):
+        """
+        Get the quantize/dequantize op index for further inserting.
+        Args:
+            The program desc.
+        Returns:
+            Two lists contains the quantize op and dequantize op index information.
+        """
+        quantize_op_index = []
+        dequantize_op_index = []
+
+        if len(self._conv_op_index) < 2:
+            return [], []
+
+        for index, value in enumerate(self._conv_op_index):
```

Comment:
For the first answer:  Correct me if I'm wrong, but isn't conv_op_index[index+1] where quantization should be inserted when index == 0, instead of conv_op_index[index+1]-1, to be right before whatever was at that index? 
Usually if we want to insert something before the first element of array (i=0) we would insert at i=0 and the previous element gets pushed to i=1, so that's what I don't get here. Is it working differently here?

For the second answer: Do you always expect pooling after last convolution?

Reply:
> For the first answer: Correct me if I'm wrong, but isn't conv_op_index[index+1] where quantization should be inserted when index == 0, instead of conv_op_index[index+1]-1, to be right before whatever was at that index?
Usually if we want to insert something before the first element of array (i=0) we would insert at i=0 and the previous element gets pushed to i=1, so that's what I don't get here. Is it working differently here?
It's a trade-off `offset` here, we adjust the offset in the following __update_program to avoid much more `-1` indexing. I'll restore the original implementation.

> For the second answer: Do you always expect pooling after last convolution?
Of coz no.

Reply:
Could you explain it in more detail please?

Reply:
Regarding this:
>For example, conv(int8)--->pooling(int8) where conv is the last conv op, if we inserted the dequantize after the conv, we can't levarage the pooling op which already enabled the int8 supporting. So we just insert the dequantize after the pool op.

and the fact that you said that you don't always expect pooling after the last convolution, then how can I know if this should always work? How do you know that you can safely insert the dequantize operator after the operator following the last convolution?

Reply:
Maybe u misunderstanding  my previous comment and vice versa:).
what i said that i don't expect pooling after the last convolution means that the code doesn't aware there's pooling or other non-conv-int8 op following by last conv int8 op. So there're two cases here:

1.  there's int8-enabled-op following by last conv op, such as pooling, we'll add the dequantize op after the pool2d. 

2. there's no int8-enabled-op following by last conv op, we'll set the last conv op attribute `force_fp32_output` to True without dequantize op insertion.